csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -256,350 +173,224 @@ public void
+                // If we already have a locator assigned, ensure the given locator has the same identity, otherwise
+                // ignore it.
+                if (_locator != null && !locator.Identity.Category.Equals(_locator.Identity.Category))
                 {
                     if (!_warned)
                     {
                         _warned = true; // Only warn once
 
-                        locator.Communicator.Logger.Warning(
-                            "received Ice locator with different instance name:\n" +
-                            "using = `" + _locator.Identity.Category + "'\n" +
-                            "received = `" + locator.Identity.Category + "'\n" +
-                            "This is typically the case if multiple Ice locators with different " +
-                            "instance names are deployed and the property `IceLocatorDiscovery.InstanceName'" +
-                            "is not set.");
+                        var s = new StringBuilder();
+                        s.Append("received Ice locator with different instance name:\n")
+                         .Append("using = `").Append(_locator.Identity.Category).Append("'\n")
+                         .Append("received = `").Append(locator.Identity.Category).Append("'\n")
+                         .Append("This is typically the case if multiple Ice locators with different ")
+                         .Append("instance names are deployed and the property `IceLocatorDiscovery.InstanceName' ")
+                         .Append("is not set.");
+                        locator.Communicator.Logger.Warning(s.ToString());
                     }
                     return;
                 }
 
-                if (_pending) // No need to continue, we found a locator
-                {
-                    _timer.Cancel(this);
-                    _pendingRetryCount = 0;
-                    _pending = false;
-                }
-
                 if (_traceLevel > 0)
                 {
                     var s = new StringBuilder("locator lookup succeeded:\nlocator = ");
                     s.Append(locator);
-                    if (_instanceName.Length == 0)
+                    if (_instanceName.Length > 0)
                     {
                         s.Append("\ninstance name = ").Append(_instanceName);
                     }
                     _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
                 }
 
-                ILocatorPrx? l = null;
-                if (_pendingRequests.Count == 0)
+                if (_locator == null)
                 {
-                    _locators.TryGetValue(locator.Identity.Category, out _locator);
+                    if (_completionSource == null)
+                    {
+                        // Ignore reply to request that already timeout.
+                        return;
+                    }
```

Comment:
I don't think the previous code ignored the reply here. We can still set the _locator data member even if there was a timeout. The next request will use this locator.

Reply:
right I miss that, I will fix it